internal/repositories/userRepositoryGet.go

```diff
@@ -0,0 +1,116 @@
+package repositories
+
+import (
+	"context"
+	"errors"
+	"github.com/loukaspe/platform-go-challenge/internal/core/domain"
+	apierrors "github.com/loukaspe/platform-go-challenge/pkg/errors"
+	"gorm.io/gorm"
+	"net/http"
+	"strconv"
+	"sync"
+)
+
+func (repo *UserRepository) GetUser(
+	ctx context.Context,
+	uid uint32,
+) (*domain.User, error) {
+	var err error
+	var modelUser *User
+	var mutex sync.Mutex
+	var wg sync.WaitGroup
+
+	err = repo.db.WithContext(ctx).
+		Preload("FavouriteCharts").
+		Preload("FavouriteAudiences").
+		Preload("FavouriteInsights").
+		Model(User{}).
+		Where("id = ?", uid).
+		Take(&modelUser).Error
+
+	if err == gorm.ErrRecordNotFound {
+		return &domain.User{}, apierrors.UserNotFoundErrorWrapper{
+			ReturnedStatusCode: http.StatusNotFound,
+			OriginalError:      errors.New("userId " + strconv.Itoa(int(uid)) + " not found"),
+		}
+	}
+
+	if err != nil {
+		return &domain.User{}, err
+	}
+
+	var favouriteAssets []domain.Asset
+
+	wg.Add(1)
+	go func(
+		charts *[]Chart,
+		favouriteAssets *[]domain.Asset,
+		mutex *sync.Mutex,
+	) {
+		defer wg.Done()
+
+		for _, chart := range modelUser.FavouriteCharts {
+			mutex.Lock()
+			*favouriteAssets = append(*favouriteAssets, domain.Chart{
+				Title:       chart.Title,
+				XAxisTitle:  chart.XTitle,
+				YAxisTitle:  chart.YTitle,
+				Data:        chart.Data,
+				Description: chart.Description,
+				AssetType:   "chart",
+			})
+			mutex.Unlock()
+		}
+	}(&modelUser.FavouriteCharts, &favouriteAssets, &mutex)
```

Comment:
Good job on making this piece of code concurrent 👍  One thing you might have heard is that parallelizing a piece of code doesn't guarantee performance speed up. In fact, the opposite may happen. What are the reasons for that? Is there any chance your "parallel" implementation is slower than the "sequential" equivalent? If you had the time, how would you benchmark it?

Reply:
Yes, indeed parallelizing could end up in slower speed. First reason could be bad synchonization in the code between go routines that end up in slower performance. But even if the synchronizing is done correctly by the engineer, synchronizing could create an overhead and cause goroutines to wait for each other. Secondly, a not balanced distribution of load could result in bad performance. For example in our problem, to have 100 assets with 98 charts, 1 insight and 1 audience.

I would test it in two ways. Firstly, write benchmarks unit tests for this function, using Go's `func BenchmarkXxx(*testing.B)` tests.
Secondly with actual integration tests and/or UAT to check how it performs. If the performance is acceptable, then we're OK. If not, then I would do the same tests for both solutions and choose the better one.